juju/terraform-provider-juju
# Support for credential management (#142)

Fix #142

Support for credential management:
- [x] [juju add-credentials](https://juju.is/docs/olm/add-credentials) 
- [x] [juju update-credential](https://juju.is/docs/olm/update-a-credential) 
- [x] [juju remove-credential](https://juju.is/docs/olm/remove-a-credential)

## QA steps

- Create two resources, model and credential, by applying the following plan:

```
resource "juju_credential" "mycredential" {
  name = "mycredential"
  auth_type = "oauth2"
  cloud {
    name   = "microk8s"
  }
}

resource "juju_model" "mymodel" {
  name = "mymodel"

  cloud {
    name   = "microk8s"
  }
  credential = juju_credential.mycredential.name
  config = {
    logging-config              = "<root>=INFO"
    development                 = true
    no-proxy                    = "jujucharms.com"
    update-status-hook-interval = "5m"
  }
}
```

- Check if "mycredential" was added only to the controller.
```
$ juju credentials

Controller Credentials:
Cloud     Credentials
microk8s  microk8s, mycredential

Client Credentials:
Cloud      Credentials
localhost  localhost*
microk8s   microk8s*
```
- Check if "mycredential" is the credential used by the model "mymodel".
```
$ juju show-model mymodel
mymodel:
  name: admin/mymodel
  short-name: mymodel
  model-uuid: d0800214-5112-48f1-86fd-a3fdc6b320f0
  model-type: caas
  controller-uuid: 0720db6d-b3ff-401d-868a-958fc119808d
  controller-name: test-controller
  is-controller: false
  owner: admin
  cloud: microk8s
  region: localhost
  type: kubernetes
  life: alive
  status:
    current: available
    since: 26 seconds ago
  users:
    admin:
      display-name: admin
      access: admin
      last-connection: never connected
  sla: unsupported
  agent-version: 2.9.38
  credential:
    name: mycredential
    owner: admin
    cloud: microk8s
    validity-check: valid
  supported-features:
  - name: juju
    description: the version of Juju used by the model
    version: 2.9.38
  - name: k8s-api
    description: the Kubernetes API lets charms query and manipulate the state of
      API objects in a Kubernetes cluster
    version: 1.24.0
```
- Optional: Try to change the auth_type of "mycredential" to "invalid".
```
resource "juju_credential" "mycredential" {
  name = "mycredential"
  auth_type = "invalid"
  cloud {
    name   = "microk8s"
  }
}
```
```
Error: supported auth-types ["certificate" "clientcertificate" "oauth2" "oauth2withcert" "userpass"], "invalid" not supported
```

- Optional: Try to remove the resource "mycredential". You'll get the same behavior as the CLI not allowing the removal of a credential in use:
```
resource "juju_model" "mymodel" {
  name = "mymodel"

  cloud {
    name   = "microk8s"
  }
  credential = "microk8s"
  config = {
    logging-config              = "<root>=INFO"
    development                 = true
    no-proxy                    = "jujucharms.com"
    update-status-hook-interval = "5m"
  }
}
```
```
Error: cannot revoke credential cloudcred-microk8s_admin_mycredential: it is still used by 1 model
```
- To solve this, first change the model's credential.
```
resource "juju_credential" "mycredential" {
  name = "mycredential"
  auth_type = "oauth2"
  cloud {
    name   = "microk8s"
  }
}

resource "juju_model" "mymodel" {
  name = "mymodel"

  cloud {
    name   = "microk8s"
  }
  credential = "microk8s"
  config = {
    logging-config              = "<root>=INFO"
    development                 = true
    no-proxy                    = "jujucharms.com"
    update-status-hook-interval = "5m"
  }
}
```
- Check if "microk8s" is the credential used by the model "mymodel".
```
$ juju show-model mymodel
mymodel:
  name: admin/mymodel
  short-name: mymodel
  model-uuid: d0800214-5112-48f1-86fd-a3fdc6b320f0
  model-type: caas
  controller-uuid: 0720db6d-b3ff-401d-868a-958fc119808d
  controller-name: test-controller
  is-controller: false
  owner: admin
  cloud: microk8s
  region: localhost
  type: kubernetes
  life: alive
  status:
    current: available
    since: 26 seconds ago
  users:
    admin:
      display-name: admin
      access: admin
      last-connection: never connected
  sla: unsupported
  agent-version: 2.9.38
  credential:
    name: microk8s
    owner: admin
    cloud: microk8s
    validity-check: valid
  supported-features:
  - name: juju
    description: the version of Juju used by the model
    version: 2.9.38
  - name: k8s-api
    description: the Kubernetes API lets charms query and manipulate the state of
      API objects in a Kubernetes cluster
    version: 1.24.0
```
- Remove the credential
```
resource "juju_model" "mymodel" {
  name = "mymodel"

  cloud {
    name   = "microk8s"
  }
  credential = "microk8s"
  config = {
    logging-config              = "<root>=INFO"
    development                 = true
    no-proxy                    = "jujucharms.com"
    update-status-hook-interval = "5m"
  }
}
```
- Check if "mycredential" was removed.
```
$ juju credentials

Controller Credentials:
Cloud     Credentials
microk8s  microk8s

Client Credentials:
Cloud      Credentials
localhost  localhost*
microk8s   microk8s*
```
